File: bots/random.cpp
#include <cstdlib>
#include <ctime>
#include <iostream>
#include <vector>

using namespace std;

struct Card {
  char rank, suit;

  bool empty() {
    return rank == 'X' || suit == 'X';
  }
};

struct Trick {
  Card cards[4];
  int startingPlayer;
  char suit;
};

struct Input {
  int currentPlayer;
  vector<Card> cardsInHand;
  int trumpPlayer;
  Card trumpCard;
  Trick currentTrick;
  vector<Trick> previousTricks;
  int points[2];
};

void readInput(Input& in) {
  cin >> in.currentPlayer;
  unsigned int NC; cin >> NC;
  in.cardsInHand.resize(NC);
  for (Card& c : in.cardsInHand) {
    cin >> c.rank >> c.suit;
  }
  cin >> in.trumpPlayer;
  cin >> in.trumpCard.rank >> in.trumpCard.suit;
  cin >> in.currentTrick.startingPlayer;
  for (Card& card : in.currentTrick.cards) {
    cin >> card.rank;
    if (card.rank != 'X') {
      cin >> card.suit;
    } else {
      card.suit = 'X';
    }
  }
  cin >> in.currentTrick.suit;
  unsigned int P; cin >> P;
  in.previousTricks.resize(P);
  for (Trick& t : in.previousTricks) {
    cin >> t.startingPlayer;
    for (Card& card : t.cards)
      cin >> card.rank >> card.suit;
    t.suit = t.cards[t.startingPlayer].suit;
  }
  cin >> in.points[0] >> in.points[1];
}

Card play(Input& in) {
  srand(time(NULL));

  vector<Card> validCards;
  if (in.currentPlayer == in.currentTrick.startingPlayer)
    validCards = in.cardsInHand;
  else {
    for (Card& c : in.cardsInHand) {
      if (c.suit == in.currentTrick.suit)
        validCards.push_back(c);
    }
    if (validCards.empty())
      validCards = in.cardsInHand;
  }
  return validCards[rand() % validCards.size()];
}

int main() {
  Input in; readInput(in);
  Card c = play(in);
  cout << c.rank << c.suit << endl;
}

File: bots/random.py
import sys
import random


class Card:
    def __init__(self, card):
        if card[0] != 'X':
            self.rank = card[0]
            self.suit = card[1]
        else:
            self.rank = ''
            self.suit = 'X'


class Hand:
    def __init__(self, cards_line):
        cards_list = cards_line.split(' ')[1:]
        self.cards = [Card(card) for card in cards_list]


class Trick:
    def __init__(self, trick_data):
        if not isinstance(trick_data, list):
            trick_data = trick_data.split(' ')
        self.cards = [Card(card) for card in trick_data[1:]]
        self.starting_player = int(trick_data[0])
        self.suit = self.cards[self.starting_player].suit


class Input:
    def __init__(self, input_data):
        self.player_number = int(input_data[0])
        self.hand = Hand(input_data[1])
        self.trump_player = int(input_data[2])
        self.trump_card = Card(input_data[3])
        self.current_trick = Trick(input_data[4])
        self.current_suit = input_data[5].strip()

        previous_tricks_data = input_data[6].split(' ')
        tricks_count = int(previous_tricks_data[0])
        self.previous_tricks = []
        for i in range(tricks_count):
            start = i*5+1
            trick = Trick(previous_tricks_data[start:start+5])
            self.previous_tricks.append(trick)

        point_data = input_data[7].split(' ')
        self.points = [int(point_data[0]), int(point_data[1])]


def play(input):
    for card in input.hand.cards:
        if card.suit == input.current_suit:
            return card
    return random.choice(input.hand.cards)


if __name__ == '__main__':
    input = Input(sys.stdin.readlines())
    card = play(input)
    print(card.rank + card.suit)
